content/blog/mcp-with-wasmcp.md

```diff
@@ -0,0 +1,470 @@
+
+All LLMs depend on calling external [functions](https://gorilla.cs.berkeley.edu/leaderboard.html), also called tools, to interact with the outside world beyond the prompt and to perform deterministic actions. Just like you might use a calculator to accurately crunch numbers, or a web browser to explore the internet, an LLM might use its own calculator and HTTP fetch tools in the same way. Even basic capabilities like reading a file from disk are implemented via tools.
+
+Without tools a language model is like someone sitting in an empty, windowless box with only their memories from an array of random encyclopedias, books, and other training data to pull from. Our interactions with them are something along the lines of: A human slips a question written on a piece of paper under the door for the model to read, and the model slips back a response using only their prior knowledge and imagination.
+
+That's a long way from the promise of autonomous systems that understand and act on the world in realtime, let alone transform it.
+
+Our first thought might be to implement a simple HTTP fetch tool for our target model. Now that model can search the internet in a loop against user queries and, voilà, we have an *agent*. Fresh data and the means of interacting with the current state of the world become available.
+
+That windowless box gets a desktop with a browser.
+
+Problem solved, right? Not quite.
+
+### Communication Hurdles
+
+**Problem 1**: The internal representation of tools is not standard across models. In other words, their hands look different. How do we build a hammer that each of them can grip?
+
+We’d need to write a new implementation of each tool for OpenAI’s GPT models, another for the Claude family, another for Gemini, etc. So the number of total tool implementations is  `M x N`, where `M` is the number of models and `N` is the number of unique tools.
+
+[AI SDKs](https://ai-sdk.dev/docs/ai-sdk-core/tools-and-tool-calling) can alleviate this problem for a given programming language by implementing tool calling support for multiple models and exposing a common interface for tools over them.
+
+**Problem 2**: Tool calling implemented by an AI SDK couples tool instances to an application's runtime. Tools must run alongside the same code that calls inference to implement the agent's loop. We cannot take one application's tools and call them from an external process.
+
+We want to implement a given tool only once and make it discoverable and accessible dynamically for any AI application, potentially across the network, at scale.
+
+The [Fundamental Theorem of Software Engineering](https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering) states:
+
+> We can solve any problem by introducing an extra level of [indirection](https://en.wikipedia.org/wiki/Indirection).
```

Comment:
Nit: In markdown, this seems to hide this quote rather than highlight it. I'd just do a classic quote.
```suggestion
The [Fundamental Theorem of Software Engineering](https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering) states: "We can solve any problem by introducing an extra level of [indirection](https://en.wikipedia.org/wiki/Indirection)."
```

Reply:
That seems like a bug in our stylesheets (or something)?  `>` should render as blockquote.  Or am I misunderstanding your comment Kate?